File: bench/Reference/queens.hs
safe :: Int -> [Int] -> Int -> Bool
safe _ [] _ = True
safe queen (q:rest) col
  | queen == q = False
  | abs (queen - q) == col = False
  | otherwise = safe queen rest (col + 1)

solve :: Int -> Int -> [Int] -> Int
solve n row placed
  | row > n = 1
  | otherwise = tryCol 1
  where
    tryCol col
      | col > n = 0
      | safe col placed 1 = solve n (row + 1) (col : placed) + tryCol (col + 1)
      | otherwise = tryCol (col + 1)

main :: IO ()
main = print (solve 10 1 [])

File: bench/Reference/queens.erl
#!/usr/bin/env escript
%% 10-queens count

main(_) ->
    io:format("~p~n", [solve(10, 1, [])]).

solve(N, Row, _Placed) when Row > N -> 1;
solve(N, Row, Placed) -> try_col(N, Row, Placed, 1).

try_col(N, _Row, _Placed, Col) when Col > N -> 0;
try_col(N, Row, Placed, Col) ->
    With = case safe(Col, Placed, 1) of
        true  -> solve(N, Row + 1, [Col | Placed]);
        false -> 0
    end,
    With + try_col(N, Row, Placed, Col + 1).

safe(_Q, [], _C) -> true;
safe(Q, [X | _], _C) when Q =:= X -> false;
safe(Q, [X | _], C) when abs(Q - X) =:= C -> false;
safe(Q, [_ | Rest], C) -> safe(Q, Rest, C + 1).
